There are a few issues with loading the data, some improperly formatted rows, if choosing to ignore the bad rows it contains 9373 rows in total from a CSV that had 10688 rows, rather large amount due to mismatch between number of items on a line and the number of columns, due to line breaks in the descriptions only 17 of the lines were badly formatted.

A cause for concern there is tweet data for 9662 users but the user table only contains 9380 users.

In [1]:
from gender_detector.gender_detector import GenderDetector as gd
import numpy as np
import pandas as pd
import re, json
import separateGUsers as sgu

twitter_Users = pd.read_csv("usersTwitter.csv", sep=",", encoding="utf-8", index_col=False, usecols=range(20))
#twitter_Users.head()
#twitter_Users
twitter_Users.shape

(9380, 20)

# Check For Duplicate Users and Tweets

In [38]:
twitter_Users.duplicated().value_counts()

False    8610
True      770
dtype: int64

770 duplicate rows in the user frame, will need to be dropped

Take list of google users, compare to twitter users to ensure one to one correlation and removing pages and keeping people. Select columns that are of interest and relevant to the model and write to file

__Don't run cell below more than once or if you do re-run the cells below this one__

In [15]:
users = sgu.read_dict_pairs()

twitter_users = pd.read_csv("usersTwitter.csv", sep=",", encoding="utf-8", index_col=False, usecols=range(20))
twFix = twitter_users[twitter_users.screen_name.isin(users.values())].drop_duplicates()
twFix = twFix[["id_str", "name", "screen_name", "location", "description", "url", "followers_count", "friends_count", "listed_count", "statuses_count", "lang"]]
twFix.set_index("id_str").to_csv("usersTrueTwitter.csv", index_label="id_str", encoding="utf-8")


Read in filtered twitter users

In [3]:
users = pd.read_csv("usersTrueTwitter.csv", sep=",", index_col="id_str", encoding="utf-8", engine='python')
users.shape

(7004, 7)

Testing gender guesser. US was selected because it is likely to include spanish names as well

In [17]:
test = gd("us")
test.guess("Emilia Miller")
name = users.iloc[0,0]
fname = name.split()[0]
test.guess(fname) #Justin
test.guess("the")

'male'

Try out regular expression match first token of only characters as some punctuation or numbers is mixed with it.

In [18]:
reg = re.compile("[a-z]*", re.IGNORECASE)
match = reg.match("Ber.t J. Gives, Jr.")


Assign genders to the twitter users using the package. Haven't created checks for accuracy but should. The name string is tokenised and the first token is match to the regexp and guessed. Sometimes the thing throws KeyErrors (disallowed chars) and IndexErrors (empty string).

__Takes a while to run so please save yourself the pain and load in the file I've already prepped. If you overwrite re-run this cell__

In [19]:
genderlist = []
gend = gd("us")
reg = re.compile("[a-z]*", re.IGNORECASE)
for name in users.name:
    try:
        if name is None:
            raise KeyError
        name = name.split()[0] #Take the first name prior to whitespace
        name = reg.match(name).group()
        gender = gend.guess(name)
        genderlist.append(gender)
    except (KeyError, IndexError):
        genderlist.append("unknown")
print("Done")

Done


Write the dataframe to a file, too time consuming to run all these cells whenever I want to start

In [20]:
users['gender'] = genderlist
users.to_csv("usersTrueTwitter.csv", index_label="id_str", encoding="utf-8")

Read in both dataframes for the user data __START HERE__

In [11]:
usersG = pd.read_csv("usersGoogle.csv", sep=",", index_col="id", quotechar="\"", encoding="utf-8", engine="c")
usersT = pd.read_csv("usersTrueTwitter.csv", sep=",", index_col="id_str", encoding="utf-8", engine='python')

Encode the Gender Column for both dataframes, dictionary initialised beforehand to see one-hot encoding

In [12]:
encDict = {"unknown":0,"none":0,"other":0,"male":1,"female":2}
usersT['gender_enc'] = usersT.gender.apply(lambda x: encDict[str(x).lower()])
usersG['gender_enc'] = usersG.gender.apply(lambda x: encDict[str(x).lower()])

In [13]:
usersT = usersT[[ "name", "screen_name", "location", "description", "friends_count", "gender", "gender_enc"]]
usersT

name      screen_name  \
id_str                                                           
235020718                      Justin Esparza        JooseBoks   
22450750                   Bert J. Gives, Jr.         bkbert23   
8011132                                mike95           mike95   
96217215                         Elidier Moya         elidier_   
98688710                         Ashutosh Jha            ajha1   
139496318                       Ignacio Perez          rizos46   
227128132                     Bárbara Botello  Barbara_Botello   
64238125                        Emilia Miller    EmiliaMaria76   
17189517                             persuede         persuede   
15921118                           Ren Newman          ren_new   
17108385                      The MeXicaN 420    themexican420   
64439230                         Rogue Theory      RogueTheory   
15181835                     Freddy Roosevelt  FreddyRoosevelt   
12846942                       Rowan Evenstar           zenrhe   
55010528                           Rich Ellis         ellisium   
29579556                          Mike Taylor          mctaylr   
100559927               Carlos Alejandri Peña         yaraguan   
65268529                       angga ramadhan      angga_rmdhn   
116520629                        Tinta Social      tintasocial   
63695781                       Raisa Andriana        raisa6690   
223404972                    Daily Image Buzz   DailyImageBuzz   
928179097                          Juan Mejia           Meje95   
262061502                             M Rifai    mohammedrivai   
56953039                          Daniel Hare         DaniHare   
327781075                   BUTLER BY THE SEA  _butlerbythesea   
17198308                         armandogomez     armandogomez   
86519988                              Irlinza     irlinzafarah   
145685200                     Natália Dayrell       natdayrell   
100284857                                 L N      KittieL2782   
260817994                  David-Ray Tremblay        davidrayT   
...                                       ...              ...   
21790658            The British Horse Society     BritishHorse   
270926266                  Juan Bauty ilustra    jbautyoficial   
26013179                           Peter Wong      peteypatter   
97253662                                Kasia           Ka_xia   
45026079                                   Em  AmazingFlyingEm   
55580450                           Naeem Alvi           NPAlvi   
15509357                        Rodrigo Jerry     rodrigojerry   
357756111                  Federico Di Fresco     FedeDiFresco   
52469009                       Russell Loving    RussellLoving   
323230615                    Bader B Al Harbi     BaDrAlHeLaLi   
120461240                           Domluiz@o        DomLuizao   
2174490787                        shower head       ShowerInfo   
164979747                      Alastair Slade   alastaireslade   
346952355                      Saleh M. Alowa       Saleh1411z   
34128158                           Jenni Paul        JenniPaul   
220364998                Julius Doroskevicius   Designbyjulius   
25699762                    Harry C Katopodis   HarryKatopodis   
143791240                       jamal mahamed      jamaal77777   
179794726                      Dusty Clement     Dusty_Clement   
261030700                  sebastian anabalon    wainsebastian   
584693106                              Dolsha   TheNobleKnight   
222987652                           KC Lashes   lecachetlounge   
193125571                  Julie A. Ketterman   julieketterman   
10946072                         Julian Klein          juliank   
863355533994143744            Дмитрий Куликов        hakudzero   
34674813                          Dave Kramer        kramersrf   
115364479                      Andrés Catalán      catalandres   
31886123                 Bob&Heather Langston    ArtByLangston   
86793577       

In [14]:
test1 = "{'familyName': 'R.', 'givenName': 'Rebecca'}"
test2 = "{'familyName': 'O'Neill', 'givenName': 'Michael'}"

regexp = re.sub(r'{\'familyName\': \'(.*)\', \'givenName\': \'(.*)\'}',r'{"familyName": "\1", "givenName": "\2"}',test1)

Find the full name of the person from extracting the name dictionary and concatenating first and last names. Initially thought json loads would be sufficient but names with apostrophes cause issues when formating for json strings so a regular expression was used instead

In [15]:
def gather_name(name):
    name = re.sub(r'{\'familyName\': \'(.*)\', \'givenName\': \'(.*)\'}',r'{"familyName": "\1", "givenName": "\2"}',name)
    name = json.loads(name)
    return str(name["givenName"] + " " + name["familyName"])


usersG["fullname"] = usersG.name.apply(lambda x:gather_name(x))

Find location from locations dictionary similar problem as previously but instead the location listed as primary was chosen, a user could have multiple locations which werent primary so it was substituted for none instead. If multiple location which were listed as primary (which there should not be) the first match was chosen.

In [16]:
def gather_location(location):
    pat = re.compile("'value': '(.*)', 'prim", re.IGNORECASE)
    if location:
        location = re.search(pat, location)
        if location:
            return location.group(1)
    return None
        
usersG["primary_location"] = usersG.placesLived.apply(lambda x: gather_location(x))

Append occupation and tagline to match the Twitter description equivalent, people tend to also post their occupation there

In [17]:
occ = usersG.occupation
tag = usersG.tagline

desc = []

for job, line in zip(occ, tag):
    if job == "None":
        job = ""
    if line == "None":
        line = ""
    desc.append(str(line + " " + job).strip())
usersG['description'] = desc

In [18]:
usersG = usersG[['fullname', 'displayName', 'primary_location', 'description', 'circledByCount', 'gender', "gender_enc"]]
usersG

fullname  \
id                                               
104288964437721273491          Johnny Oliveira   
100301662419920433743           Justin Esparza   
115233306327496577248           Bert Gives, Jr   
114450247448428643447          Michael Olivero   
106057007019277820892             Elidier Moya   
115538752986453113378             Ashutosh Jha   
108359110716028319267    ignacio perez alvarez   
117585428399967312212          Bárbara Botello   
112929377664095396761            Emilia Miller   
100502806126131626792                per suede   
113976667959203378689               Ren Newman   
116740965093909274586        Joseph TheMeXicaN   
118152866799626820488           Marcus Lambert   
100255120107004491447                Kim Mason   
112091801271920083184         Freddy Roosevelt   
116069220457248636438           Rowan Evenstar   
114585659578085774680               Rich Ellis   
105218321052691174487              Mike Taylor   
103320849848035973570    Carlos Alejandri Peña   
104999784497781549142           Angga Ramadhan   
103277343880765119232  Felipe pereira da silva   
112949810091403026543             Tinta Social   
112849973790998332988           Raisa Andriana   
116517034830139229914         Daily Image Buzz   
110234035865091632953                Mejia Lkp   
108426846048332320933                  M Rifai   
110211016089328289280           Yasuko Ikegami   
110192233647107444825              Daniel Hare   
114949087075925464316           Scott Cansdell   
105045867837097616877            Armando Gomez   
...                                        ...   
111041273287874791161               Emily Keer   
105631162554914194689               Naeem Alvi   
102317163369188680281            Rodrigo Jerry   
110461231329614771838       Federico Di Fresco   
117989090104296344397           Russell Loving   
106154423834861027391           BaDr Al HeLaLi   
118017659495777493067                Luiz Gama   
108325257053453452300            TAKE A SHOWER   
116068929718715118792           Alastair Slade   
112157412724215731992           Saleh M. Alowa   
113665062457287670639         Project PeacePal   
115785027448099756630               Jenni Paul   
100327555683961109682     Julius Doroskevicius   
117777130232095923626            Larry Kuehner   
112280250490980073383          Harry Katopodis   
100509467460696324025            Jamal Mahamed   
116613726653221664471            Dusty Clement   
103029057458261501702          sebastian perez   
110133348069637582755           Michael Hurley   
109609108757245537962              Tasha Socin   
118135680861666022405          Julie Ketterman   
100935077982279403907             Julian Klein   
112708269319876235025               haku dzero   
102647525493626642487              Dave Kramer   
112881479665900679552           Andrés Catalán   
112161508125849947904         Heather Langston   
109420270642726200915        Misti Hope Wudtke   
107017704360703857839       Marc Caymel Romero   
111123839983171581398          AnalistaTI Blog   
107984737937791577439            wilbert lopez   

                                            displayName  \
id                                                        
104288964437721273491                   Johnny Oliveira   
100301662419920433743                    Justin Esparza   
115233306327496577248                    Bert Gives, Jr   
114450247448428643447                   Michael Olivero   
106057007019277820892                      Elidier Moya   
115538752986453113378                      Ashutosh Jha   
108359110716028319267             ignacio perez alvarez   
117585428399967312212                   Bárbara Botello   
112929377664095396761                     Emilia Miller   
100502806126131626792                         per suede   
113976667959203378689                        Ren Newman   
116740965093909274586                 Joseph TheMeXicaN   
118152866799626820488                    Marcus Lambert   
10

Write to file one last time

In [19]:
usersT.to_csv("usersTrueTwitter.csv", index_label="id_str", encoding="utf-8")
usersG.to_csv("usersGoogle.csv", index_label="id", encoding="utf-8")